runtime/error.c

```diff
@@ -1,12 +1,128 @@
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+
 #include "balrt.h"
+#include "third-party/libbacktrace/backtrace.h"
+
+#define INITIAL_PC_COUNT 1
+#define MAX_PC_COUNT 256
+#define SKIP_FROM_END 1 // This number counts from the last stack frame
+#define THREAD 0
+#define SUCCESS 0
+#define FAIL 1
+
+struct backtrace_state *state = NULL;
+
+typedef struct {
+    uint32_t nPCs;
+    uint32_t szPCs;
+    bool errorOccurred;
+    PC *pcs;
+} PCVector;
+
+static void setBackTraceSimpleInternalError(void *data, const char *msg, int errnum);
+static void setBackTracePCInfoInternalError(void *data, const char *msg, int errnum);
+static int printBacktraceLine(void *data, PC pc, const char *filename, int lineno, const char *function);
+static int updatePCVector(void *data, PC pc);
+static void getPCs(PCVector *pcVector);
 
 TaggedPtr _bal_error_construct(TaggedPtr message, int64_t lineNumber) {
-    ErrorPtr ep = _bal_alloc(sizeof(struct Error));
+    PCVector pcVector = {0, INITIAL_PC_COUNT, false};
+    getPCs(&pcVector);
+
+    uint32_t nPCs = pcVector.nPCs;
+    uint64_t errStructSize = sizeof(struct Error) + sizeof(PC) * nPCs;
+    ErrorPtr ep = _bal_alloc(errStructSize);
     ep->message = message;
     ep->lineNumber = lineNumber;
+    ep->internalErrorOccured = pcVector.errorOccurred;
+    ep->nPCs = nPCs;
+
+    PC *pcs = pcVector.pcs;
+    memcpy(ep->pcs, pcs, sizeof(PC) * nPCs);
+    free(pcs);
+
     return ptrAddFlags(ep, (uint64_t)TAG_ERROR << TAG_SHIFT);
 }
 
+static void getPCs(PCVector *pcVector) {
+    if (state == NULL) {
+        state = backtrace_create_state(NULL, THREAD, NULL, NULL);
+        if (state == NULL) {
+            pcVector->errorOccurred = true;
+            return;
+        }
+    }
+
+    void *p = malloc(sizeof(PC) * INITIAL_PC_COUNT);
```

Comment:
It's better to do this malloc in _bal_error_construct so that the pcVector is always in a consistent state with the allocated size of pcs equal to szpcs.

Reply:
Changed